### PR TITLE
Decode bytes object from sql_ascii databases

### LINKED ISSUE #584: `postgresql_user` fails on `SQL_ASCII` encoded database
##### SUMMARY
Hello!

The `postgresql_user` module is failing on an `SQL_ASCII` encoded database when trying to check if the user's password should be changed. This is because `psycopg` parses the data returned from the database as [ascii encoded `bytes`](https://www.psycopg.org/psycopg3/docs/basic/adapt.html#strings-adaptation) (PostgreSQL treats `SQL_ASCII` as no encoding) instead of a utf-8 `str` and therefore the following regex match fails with `TypeError: cannot use a string pattern on a bytes-like object`:

https://github.com/ansible-collections/community.postgresql/blob/98dd2819dd4772331e64f95626abee7db0570f56/plugins/modules/postgresql_user.py#L422-L427

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
`postgresql_user`

##### ANSIBLE VERSION

```paste below
ansible [core 2.15.3]
  config file = /home/cwhite/projects/devops/ansible.cfg
  configured module search path = ['/home/cwhite/projects/devops/library']
  ansible python module location = /home/cwhite/.local/lib/python3.10/site-packages/ansible
  ansible collection location = /home/cwhite/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/cwhite/.local/bin/ansible
  python version = 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (/usr/bin/python3)
  jinja version = 3.0.3
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /home/cwhite/.ansible/collections/ansible_collections
Collection           Version
-------------------- -------
community.postgresql 3.2.0  

# /home/cwhite/.local/lib/python3.10/site-packages/ansible_collections
Collection           Version
-------------------- -------
community.postgresql 2.4.3  
```

##### CONFIGURATION

```paste below
CONFIG_FILE() = /home/cwhite/projects/devops/ansible.cfg
DEFAULT_EXECUTABLE(/home/cwhite/projects/devops/ansible.cfg) = /bin/bash
DEFAULT_HOST_LIST(/home/cwhite/projects/devops/ansible.cfg) = ['/home/cwhite/projects/devops/inventory/servers.yml']
DEFAULT_LOG_PATH(/home/cwhite/projects/devops/ansible.cfg) = /home/cwhite/ansible.log
DEFAULT_MODULE_PATH(/home/cwhite/projects/devops/ansible.cfg) = ['/home/cwhite/projects/devops/library']
DEFAULT_PRIVATE_KEY_FILE(/home/cwhite/projects/devops/ansible.cfg) = /home/cwhite/.ssh/id_rsa
DEFAULT_PRIVATE_ROLE_VARS(/home/cwhite/projects/devops/ansible.cfg) = True
DEFAULT_ROLES_PATH(/home/cwhite/projects/devops/ansible.cfg) = ['/home/cwhite/projects/devops/roles']
DEFAULT_VAULT_IDENTITY_LIST(/home/cwhite/projects/devops/ansible.cfg) = ['dev@admin/vault-keyring-client.py', 'stage@admin/vault-keyring-client.py', 'prod@admin/vault-keyr>
DEFAULT_VAULT_ID_MATCH(/home/cwhite/projects/devops/ansible.cfg) = true
INTERPRETER_PYTHON(/home/cwhite/projects/devops/ansible.cfg) = /usr/bin/python3
RETRY_FILES_ENABLED(/home/cwhite/projects/devops/ansible.cfg) = False
```

##### OS / ENVIRONMENT

- running against an Ubuntu 22.04 remote host
- `psycopg` 3.1.10 is installed on remote host (using `pip install "psycopg[binary]"`)

##### STEPS TO REPRODUCE

- create a SQL_ASCII encoded database
- attempt to update a users password (with password already set in database)

```yaml
- name: create database user
  become: true
  become_user: postgres
  community.postgresql.postgresql_user:
    db: '{{ database_name }}'
    name: '{{ database_user }}'
    password: '{{ database_user_password }}'
    no_password_changes: false
    encrypted: true
    role_attr_flags: LOGIN,NOSUPERUSER
    state: present
```

##### EXPECTED RESULTS

Task succeeds without errors.

##### ACTUAL RESULTS

Task fails with error mentioned above.

```paste below
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: cannot use a string pattern on a bytes-like object
fatal: [ems-dbd]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"<stdin>\", line 107, in <module>\n  File \"<stdin>\", line 99, in _ansiballz_main\n  File \"<st
din>\", line 47, in invoke_module\n  File \"/usr/lib/python3.10/runpy.py\", line 224, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.10/runpy.p
y\", line 96, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.10/runpy.py\", line 86, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_communit
y.postgresql.postgresql_user_payload_j46oj96o/ansible_community.postgresql.postgresql_user_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_user.py\", line 1054, in <module>\n
  File \"/tmp/ansible_community.postgresql.postgresql_user_payload_j46oj96o/ansible_community.postgresql.postgresql_user_payload.zip/ansible_collections/community/postgresql/plugins/modules/postgresql_user.
py\", line 993, in main\n  File \"/tmp/ansible_community.postgresql.postgresql_user_payload_j46oj96o/ansible_community.postgresql.postgresql_user_payload.zip/ansible_collections/community/postgresql/plugins
/modules/postgresql_user.py\", line 496, in user_alter\n  File \"/tmp/ansible_community.postgresql.postgresql_user_payload_j46oj96o/ansible_community.postgresql.postgresql_user_payload.zip/ansible_collectio
ns/community/postgresql/plugins/modules/postgresql_user.py\", line 427, in user_should_we_change_password\n  File \"/usr/lib/python3.10/re.py\", line 190, in match\n    return _compile(pattern, flags).match
(string)\nTypeError: cannot use a string pattern on a bytes-like object\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
```

### PR DESCRIPTION
**Pull request by CalebDW**
##### SUMMARY
Hello!

This PR fixes #584, however, I'm not sure if it's the only code that interacts with potentially encoded bytes.

Some other options are to:
- Do something like https://github.com/psycopg/psycopg/commit/ca6db5c213ad61af1f4690339248a3a447b31ef5
    - https://github.com/psycopg/psycopg/issues/503
    - https://github.com/psycopg/psycopg/issues/561
- Set a default connection parameter `client_encoding: utf8` in `postgres_common_argument_spec`

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`postgres_user`

### CODE REVIEW

**Review by hunleyd — CHANGES_REQUESTED** [review]
this looks ok to me, but i think we might want to:
> Set a default connection parameter client_encoding: utf8 in postgres_common_argument_spec

as you suggested. therefore, deferring until @Andersson007 returns from pto

**Comment by Andersson007** [comment]
@CalebDW hello, thanks for the fix!
Could you please:
- add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment)?
- can we add integration tests? i.e. create an encoded db and test everything works, the tests are located under the `tests/integration/targets` directory; how to run them is explained in the [quick start guide](https://docs.ansible.com/ansible/devel/community/create_pr_quick_start.html)

**Comment by CalebDW** [comment]
@Andersson007, sure I can do that. Do we still want to implement the changes requested by @hunleyd?

**Comment by CalebDW** [comment]
@Andersson007, tests and changelog fragment have been added. Let me know if you want to keep the fix as-is or try to solve on a more global level

**Comment by Andersson007** [comment]
> @Andersson007, sure I can do that. Do we still want to implement the changes requested by @hunleyd?

Folks, i don't know what to say really:) I can judge the code and general things like testing but i've no idea what would work best from the users' perspective. I would follow "keep it simple" and avoid any risky changes if there are more straightforward solutions.
@hunleyd if you're not sure, please engage other maintainers / contributors calling for help in Matrix / pinging them here.

maybe @betanummeric and @aleksvagachev has ideas

**Comment by aleksvagachev** [comment]
@CalebDW @Andersson007 @hunleyd I agree with the idea of "keep it simple". For specific parameters, you can use the "connect_params" field. We can try to leave it as it is for now, and if there are more issues related to this problem in the future, then we can add a parameter.

**Review by hunleyd — APPROVED** [review]

**Comment by Andersson007** [comment]
so as far as I can see people are ok to hold this on, right?
should we close the PR and the issue then?

**Comment by hunleyd** [comment]
> so as far as I can see people are ok to hold this on, right? should we close the PR and the issue then?

i _think_ people were ok w/ this targeted fix and wanted to hold on the question of the larger sweeping change.

**Review by Andersson007 — APPROVED** [review]

**Comment by Andersson007** [comment]
oh, apologies folks, working in the text-scanning mode these days..
@hunleyd thanks for clarifying!:)
approved, thanks a lot! Please merge.

**Comment by Andersson007** [comment]
@CalebDW could you please backport the change to stable-2 and stable-1 manually https://docs.ansible.com/ansible/devel/community/development_process.html#backporting-merged-prs-in-ansible-core ?

If you have no time, please let me know

**Comment by CalebDW** [comment]
@Andersson007, done, let me know if you need anything else!

**Comment by Andersson007** [comment]
> @Andersson007, done, let me know if you need anything else!

LGTM, merged, thanks!

**Pull request merged**
